### DISCUSSION TITLE
Necessity to check equality of object reference when PropertyChanged

### DISCUSSION THREAD

**Discussion by feast107** [AvaloniaUI/Avalonia#15767] [Q&A]
## Current
in `void SetValue(object?, Exception?)` of `ExpressionNode.cs` there is a reference check of binding value:
https://github.com/AvaloniaUI/Avalonia/blob/42aa77ebdcdb39195e452fe564991eb9eed46d44/src/Avalonia.Base/Data/Core/ExpressionNodes/ExpressionNode.cs#L238
which makes some differences between `Avalonia` and `WPF`.
In `WPF`, whoever trigger the `PropertyChanged` will unconditionally update the related view but `Avalonia` will check the ref of two values and reject update if they are equal.
I know in some case this check can reduce repeated execution, but what my code run worse just the opposite.

## Here is my scene:
My lib [Antelcat.I18N](https://github.com/Antelcat/I18N), supports both `WPF` & `Avalonia`, provides language bindings using user code bindings. Most of them are `MultiBinding`s which contains multiple keys of language source. In this case, update each key alone will trigger the calc many times, but 'language update' is always a 'group update' so I design a `notifier` which will trigger 'PropertyChanged' only one time when language update complete. It appears at the first place of `Bindings` in all the `MultiBinding`s and whose path is always points to the language source so I can get the source directly from `values[0]` in `IMultiValueConverter`.
But somehow it doesn't work, even if I trigger `PropertyChanged` ,`Avalonia` reject to update cause it found ref not actually changed.
I know there is some ways that can avoid this problem such as making notfier something else. But this make it to be more internal references and more codes just to fit this situation. I think the user code should be responsible for the handling of reference changes while framework code just need to focus on execution.
It's like you write code to prevent this from working and I write extra code to make it work.

**Comment by ent3m** [comment]
I recently encountered the same issue. I set up a DirectProperty but its setter never gets called when old and new value are equal. This leads to me writing some janky code like notifying a fake value change before the real one or wrapping the value inside a special class where equality check always return false.
I believe that the choice to perform the equality check should be in the hands of the developer and not the framework, especially when it's only an optimization technique. Imo, the equality check should be performed in AvaloniaObject.SetAndRaise or AvaloniaObject.SetValue and nowhere else. Let the developer decide how to use the framework, not having to work around it.

**Reply by ent3m** [reply]
To support my point, I want to give another reason why this equality check is unnecessary. Suppose that I set the value of the AvaloniaProperty via code behind, calling FooControl.Bar = value, I get the behavior that I want. Now if I do the same thing within my ViewModel, binding to the BarProperty in XAML, then suddenly the code doesn't work anymore. The very same piece of code, but one that relies on the framework is broken. This shouldn't be the default behavior.

**Reply by ent3m** [reply]
The issue comes from here https://github.com/AvaloniaUI/Avalonia/blob/afa124f505855a5faed39d4fbed273754f06a506/src/Avalonia.Base/Data/Core/BindingExpression.cs#L359 and here https://github.com/AvaloniaUI/Avalonia/blob/afa124f505855a5faed39d4fbed273754f06a506/src/Avalonia.Base/Data/Core/BindingExpression.cs#L514 possibly here as well?https://github.com/AvaloniaUI/Avalonia/blob/afa124f505855a5faed39d4fbed273754f06a506/src/Avalonia.Base/Data/Core/UntypedBindingExpressionBase.cs#L428 Please take a look, @grokys ?
